tensorflow_recommenders_addons/dynamic_embedding/core/kernels/lookup_impl/lookup_table_op_cpu.h

```diff
@@ -238,6 +247,80 @@ class TableWrapperOptimized final : public TableWrapperBase<K, V> {
     return Status::OK();
   }
 
+  Status save_to_hdfs(OpKernelContext* ctx, int64 value_dim,
+                      const string& filepath,
+                      const size_t buffer_size) override {
+    size_t dim = static_cast<size_t>(value_dim);
+    auto lt = table_->lock_table();
+
+    HadoopFileSystem hdfs;
+    std::unique_ptr<WritableFile> writer;
+    const string tmp_file = filepath + ".tmp";
+    TF_RETURN_IF_ERROR(hdfs.NewWritableFile(tmp_file, &writer));
+
+    uint32 record_len = sizeof(K) + sizeof(V) * dim;
+    uint64 pos = 0;
+    uint8 content[buffer_size + record_len];
+
+    for (auto it = lt.begin(); it != lt.end(); ++it) {
+      K k = it->first;
+      std::memcpy(content + pos, reinterpret_cast<uint8*>(&k), sizeof(K));
+
+      const auto& jt = it->second.data();
+      std::memcpy(content + pos + sizeof(K), reinterpret_cast<uint8*>(jt),
+                  sizeof(V) * dim);
+
+      pos += record_len;
+      if (pos > buffer_size) {
+        TF_RETURN_IF_ERROR(
+            writer->Append(StringPiece(reinterpret_cast<char*>(content), pos)));
+        pos = 0;
+      }
+    }
+
+    if (pos > 0) {
+      TF_RETURN_IF_ERROR(
+          writer->Append(StringPiece(reinterpret_cast<char*>(content), pos)));
+    }
+
+    TF_RETURN_IF_ERROR(writer->Close());
+    TF_RETURN_IF_ERROR(hdfs.RenameFile(tmp_file, filepath));
+    return Status::OK();
+  }
+
+  Status load_from_hdfs(OpKernelContext* ctx, int64 value_dim,
+                        const string& filepath,
+                        const size_t buffer_size) override {
+    size_t dim = static_cast<size_t>(value_dim);
+
+    HadoopFileSystem hdfs;
+    std::unique_ptr<RandomAccessFile> file;
+    TF_RETURN_IF_ERROR(hdfs.NewRandomAccessFile(filepath, &file));
+    std::unique_ptr<io::RandomAccessInputStream> input_stream(
+        new io::RandomAccessInputStream(file.get()));
+    io::BufferedInputStream reader(input_stream.get(), buffer_size);
+
+    uint64 file_size = 0;
+    TF_RETURN_IF_ERROR(hdfs.GetFileSize(filepath, &file_size));
+
+    tstring content;
+    uint64 record_len = sizeof(K) + sizeof(V) * dim;
+    uint64 i = 0;
+    while (i < file_size) {
+      TF_RETURN_IF_ERROR(reader.ReadNBytes(record_len, &content));
+      K k;
+      std::memcpy(&k, reinterpret_cast<uint8*>(content.data()), sizeof(K));
+
+      ValueType value_vec;
+      std::memcpy(value_vec.data(),
+                  reinterpret_cast<uint8*>(content.data()) + sizeof(K),
```

Comment:
Could we avoid these two memcopy here? Since the memory &content is avaialbe and the memory layout is fixed.

Reply:
Because our current underlying implementation of `ValueType` is `Array`, and ordinary arrays can only be turned into `Array` by copying. https://stackoverflow.com/questions/43236477/stl-array-from-c-array-without-copying

Reply:
Hmm. I do think it quite related to `Array` in the reference link.  Since your are memcpy a continous memory block here, maybe you could just get the address to K and address to value_vec directy:
```
K* pk = reinterpret_cast<K*>(content.data());
ValueType* pValue = reinterpret_cast<ValueType*>(content.data()) + sizeof(K));
table_->insert_or_assign(*k, *value_vec);
``` 
Could you please check again?

Reply:
accept